src/Functional.CQS.AOP.IoC.SimpleInjector.MetricsCapturing/ConventionBasedDecoratorRegistrationGatewayExtensions.cs

```diff
@@ -0,0 +1,98 @@
+		/// All handler implementations will have the universal metrics-capturing decorator applied to them.
+		/// </summary>
+		/// <typeparam name="TUniversalMetricsCapturingStrategy">The universal metrics capturing strategy type.</typeparam>
+		/// <param name="gateway">The gateway.</param>
+		/// <param name="configurationParameters">The configuration parameters.</param>
+		/// <returns></returns>
+		public static ConventionBasedDecoratorRegistrationGateway WithMetricsCapturingDecorator<TUniversalMetricsCapturingStrategy>(this ConventionBasedDecoratorRegistrationGateway gateway, MetricsCapturingModuleConfigurationParameters configurationParameters)
+			where TUniversalMetricsCapturingStrategy : class, IUniversalMetricsCapturingStrategy
+		{
+			gateway.RegisterMetricsCapturingDecoratorForIndividualQueryHandlerImplementations(configurationParameters);
+			gateway.RegisterMetricsCapturingDecoratorForIndividualCommandHandlerImplementations(configurationParameters);
+			gateway.RegisterUniversalMetricsCapturingDecorator<TUniversalMetricsCapturingStrategy>(configurationParameters);
+
+			return gateway;
+		}
+
+		private static void RegisterMetricsCapturingDecoratorForIndividualQueryHandlerImplementations(this ConventionBasedDecoratorRegistrationGateway gateway, MetricsCapturingModuleConfigurationParameters configurationParameters)
+		{
+			if (!configurationParameters.QuerySpecificMetricsCapturingDecoratorEnabled)
+				return;
+
+			var queryAndResultTypeWithMetricsCapturingStrategyDefinedCollection = new HashSet<QueryAndResultType>(gateway.AssemblyCollection
+				.SelectMany(assembly => assembly.GetTypes().Where(t => t.IsMetricsCapturingStrategyForQueryType()))
+				.Select(x => x.GetGenericParametersForQueryMetricsCapturingStrategyType())
+				.WhereSome());
+
+			bool hasMetricsCapturingStrategyDefinedForQuery(DecoratorPredicateContext c) => c.ToServiceAndImplementationType().HasMetricsCapturingStrategyDefined(queryAndResultTypeWithMetricsCapturingStrategyDefinedCollection);
+			gateway.Container.RegisterSingleton(typeof(IMetricsCapturingStrategyForQuery<,>), gateway.AssemblyCollection);
+			gateway.Container.RegisterDecorator(typeof(IQueryHandler<,>), typeof(QueryHandlerMetricsCapturingDecorator<,>), gateway.Lifestyle, hasMetricsCapturingStrategyDefinedForQuery);
+			gateway.Container.RegisterDecorator(typeof(IAsyncQueryHandler<,>), typeof(AsyncQueryHandlerMetricsCapturingDecorator<,>), gateway.Lifestyle, hasMetricsCapturingStrategyDefinedForQuery);
+		}
+
+		private static void RegisterMetricsCapturingDecoratorForIndividualCommandHandlerImplementations(this ConventionBasedDecoratorRegistrationGateway gateway, MetricsCapturingModuleConfigurationParameters configurationParameters)
+		{
+			if (!configurationParameters.CommandSpecificMetricsCapturingDecoratorEnabled)
+				return;
+
+			var commandAndErrorTypeWithCachingStrategyDefinedCollection = new HashSet<CommandAndErrorType>(gateway.AssemblyCollection
```

Comment:
This variable name should say `MetricsCapturingStrategy`, not `CachingStrategy`.

Reply:
Fixed